File: Trees/Binary Tree Preorder Traversal/Binary_Tree_Preorder_Traversal.cpp
//Problem Name: Binary Tree Preorder Traversal
//Problem Link: https://leetcode.com/problems/binary-tree-preorder-traversal/description/ 

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

 //Approach-1: Recursive
class Solution {
public:
    void dfsTraversal(TreeNode* root, vector<int>& ans) {
        if(root == NULL) return;
        ans.push_back(root->val);
        dfsTraversal(root->left, ans);
        dfsTraversal(root->right, ans);
    }

    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> ans;
        dfsTraversal(root, ans);
        return ans;
    }
};

//Approach-2: Iterative (using stack)
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> ans;
        if(root == NULL) return ans;
        stack<TreeNode*> st;
        st.push(root);
        while(!st.empty()) {
            TreeNode* node = st.top();
            st.pop();
            ans.push_back(node->val);
            if(node->right!=NULL) st.push(node->right);
            if(node->left!=NULL) st.push(node->left);
        }
        return ans;
    }
};

File: Trees/Binary Tree Preorder Traversal/Binary_Tree_Preorder_Traversal.java
//Problem Name: Binary Tree Preorder Traversal
//Problem Link: https://leetcode.com/problems/binary-tree-preorder-traversal/description/

//Approach-1: Recursive
class Solution {
    public void dfsTraversal(TreeNode root, List<Integer> ans) {
        if(root == null) return;
        ans.add(root.val);
        dfsTraversal(root.left, ans);
        dfsTraversal(root.right, ans);
    }

    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        dfsTraversal(root, ans);
        return ans;
    }
}

//Approach-2: Iterative (using Stack)
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        if(root == null) return ans;
        Stack<TreeNode> st = new Stack<>();
        st.push(root);
        while(!st.isEmpty()) {
            TreeNode node = st.pop();
            ans.add(node.val);
            if(node.right!=null) st.push(node.right);
            if(node.left!=null) st.push(node.left);
        }
        return ans;
    }
}